### PR TITLE
Support staging and local URLs

### PR DESCRIPTION
**Pull request by vitaliyreznikov**
# Description
Please include a summary of the change and which issue is fixed. Include the motivation for the changes, and comment on your PR if necessary for clarity.

Fixes: (link Jira Task)

#changelog Support staging and local URLs

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Test

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have created or modified unit tests to sufficiently cover my changes.
- [x] I have added my Asana task link and provided a changelog message if applicable.

### CODE REVIEW

File: app/code/community/Bolt/Boltpay/etc/system.xml [hunk 1]
```diff
@@ -480,6 +480,33 @@
...
+                  <sort_order>205</sort_order>
+                  <show_in_default>0</show_in_default>
+                  <show_in_website>0</show_in_website>
+                  <show_in_store>0</show_in_store>
+                </custom_api>
+                <custom_js translate="label">
+                  <label>Custom JS url</label>
+                  <config_path>payment/boltpay/custom_js</config_path>
+                  <frontend_type>text</frontend_type>
+                  <sort_order>205</sort_order>
```
Comment (adentong) [inline app/code/community/Bolt/Boltpay/etc/system.xml:496]: sort_order should be different

File: app/code/community/Bolt/Boltpay/etc/system.xml [hunk 2]
```diff
@@ -480,6 +480,33 @@
...
+                  <show_in_default>0</show_in_default>
+                  <show_in_website>0</show_in_website>
+                  <show_in_store>0</show_in_store>
+                </custom_js>
+                <custom_merchant translate="label">
+                  <label>Custom merchant url</label>
+                  <config_path>payment/boltpay/custom_merchant</config_path>
+                  <frontend_type>text</frontend_type>
+                  <sort_order>205</sort_order>
```
Comment (adentong) [inline app/code/community/Bolt/Boltpay/etc/system.xml:505]: same as above

**Review by adentong — APPROVED** [review]
Can you also check by integration tests failed?

File: tests/unit/testsuite/Bolt/Boltpay/Helper/UrlTraitTest.php [hunk 3]
```diff
@@ -49,110 +49,74 @@ protected function tearDown()
...
-            Bolt_Boltpay_TestHelper::getNonPublicProperty($this->currentMock, 'merchantUrlSandbox'),
-            $this->currentMock->getBoltMerchantUrl()
-        );
+    public function providerGetApiUrl() {
+        return [
+            [false, '', 'https://api.bolt.com/'],
+            [false, 'https://api.vitaliy.dev.bolt.me/', 'https://api.bolt.com/'],
+            [true, '', 'https://api-sandbox.bolt.com/'],
+            [true, 'https://wrong.url.com/', 'https://api-sandbox.bolt.com/'],
+            [true, 'https://api.vitaliy.dev.bolt.me/', 'https://api.vitaliy.dev.bolt.me/'],
```
Comment (roopakv) [inline tests/unit/testsuite/Bolt/Boltpay/Helper/UrlTraitTest.php:70]: Could you specifically add a test for staging

**Comment by vitaliyreznikov** [comment]
@adentong 
> Can you also check by integration tests failed?
The broken on master as well looks like somethings with ngrok settings.
I've asked QA team to take a look.

File: tests/unit/testsuite/Bolt/Boltpay/Helper/UrlTraitTest.php (hunk 3 above)
Comment (vitaliyreznikov) [reply tests/unit/testsuite/Bolt/Boltpay/Helper/UrlTraitTest.php:70]: Good catch, added

**Review by adentong — APPROVED** [review]

**Review by roopakv — APPROVED** [review]
LGTM from me. Please get someone more experienced with m1 to sign off

File: app/code/community/Bolt/Boltpay/Helper/UrlTrait.php [hunk 4]
```diff
@@ -84,9 +115,14 @@ public function getApiUrl($storeId = null)
...
     {
-        return  Mage::getStoreConfigFlag('payment/boltpay/test', $storeId) ?
-            $this->jsUrlTest :
-            $this->jsUrlProd ;
+        if (Mage::getStoreConfig('payment/boltpay/test', $storeId)) {
+            return $this->getCustomURLValueOrDefault(
+                Mage::getStoreConfig('payment/boltpay/custom_js', $storeID),
+                $this->jsUrlTest
+            );
+        } else {
```
Comment (johnmccombs1) [inline app/code/community/Bolt/Boltpay/Helper/UrlTrait.php:123]: @vitaliyreznikov  Can we use the ternary operator here? If you don't want to use it, can you remove the "else" condition and simply return $this->apiUrlProd?

**Review by johnmccombs1 — APPROVED** [review]

**Pull request merged**